pkg/image/controller/controller.go

```diff
@@ -39,126 +37,203 @@ const retryCount = 2
+//    Additionally:
+//    for kind == DockerImage import or reference underlying image, iow. exact tag (not provided means latest),
+//    for kind != DockerImage reference tag from the same or other ImageStream
+// 3. spec.DockerImageRepository not defined - import tags per its definition.
+// Current behavior of the controller is to process import as far as possible, but
+// we still want to keep backwards compatibility and retries, for that we'll return
+// error in the following cases:
+// 1. connection failure to upstream image repository
+// 2. reading tags when error is different from RepositoryNotFound or RegistryNotFound
+// 3. image retrieving when error is different from RepositoryNotFound, RegistryNotFound or ImageNotFound
+// 4. ImageStreamMapping save error
+// 5. error when marking ImageStream as imported
+
 func (c *ImportController) Next(stream *api.ImageStream) error {
 	if !needsImport(stream) {
 		return nil
 	}
-	name := stream.Spec.DockerImageRepository
-
-	ref, err := api.ParseDockerImageReference(name)
-	if err != nil {
-		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
-		util.HandleError(err)
-		return c.done(stream, err.Error(), retryCount)
-	}
-
-	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
 
+	insecure := stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
 	client := c.client
 	if client == nil {
 		client = dockerregistry.NewClient()
 	}
-	conn, err := client.Connect(ref.Registry, insecure)
-	if err != nil {
+
+	toImport, err := c.getTags(stream, client, insecure)
+	// return here, only if there is an error and nothing to import
+	if err != nil && len(toImport) == 0 {
 		return err
 	}
-	tags, err := conn.ImageTags(ref.Namespace, ref.Name)
-	switch {
-	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-		return c.done(stream, err.Error(), retryCount)
-	case err != nil:
+
+	errs := c.importTags(stream, toImport, client, insecure)
+	// one of retry-able error happened, we need to inform the RetryController
+	// the import should be retried by returning error
+	if len(errs) > 0 {
+		return kerrors.NewAggregate(errs)
+	}
+	if err != nil {
 		return err
 	}
 
-	imageToTag := make(map[string][]string)
-	for tag, image := range tags {
-		if specTag, ok := stream.Spec.Tags[tag]; ok && specTag.From != nil {
-			// spec tag is set to track another tag - do not import
+	return c.done(stream, "", retryCount)
+}
+
+// getTags returns tags from default upstream image repository and explicitly defined.
+// Returns a map of tags to be imported and an error if one occurs.
+// Tags explicitly defined will overwrite those from default upstream image repository.
+func (c *ImportController) getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, error) {
```

Comment:
nitpick: Controller is unused inside this method. It could be turned into a function.